### PR TITLE
Implement a Bindable List

### PR DESCRIPTION
**Pull request by miterosan**
Added Tests (81% coverage)

### CODE REVIEW

File: osu.Framework/Configuration/BindableList.cs [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+    public class BindableList<T> : IBindable, IList<T>, IReadOnlyList<T>
+    {
+        private static InvalidOperationException createDisabledExeption()
+            => new InvalidOperationException("Can not modify values as bindable list is disabled");
+
+        private readonly List<T> underlyingList = new List<T>();
+
+        public void Parse(object input)
+        {
+            if (input is IList<T>)
```
Comment (jorolf) [inline osu.Framework/Configuration/BindableList.cs]: wouldn't IEnumerable be a better choice?

File: osu.Framework/Configuration/BindableList.cs [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using osu.Framework.Lists;
+
+namespace osu.Framework.Configuration
+{
+    public class BindableList<T> : IBindable, IList<T>, IReadOnlyList<T>
+    {
+        private static InvalidOperationException createDisabledExeption()
```
Comment (jorolf) [inline osu.Framework/Configuration/BindableList.cs]: typo

File: osu.Framework/Configuration/BindableList.cs [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+        private readonly List<T> underlyingList = new List<T>();
+
+        public void Parse(object input)
+        {
+            if (!(input is IEnumerable<T>))
+                throw new ArgumentException($@"Could not parse provided {input.GetType()} ({input}) to {typeof(T)}.");
+            Clear();
+            underlyingList.AddRange((IList<T>)input);
+
+            throw new ArgumentException($@"Could not parse provided {input.GetType()} ({input}) to {typeof(T)}.");
```
Comment (jorolf) [inline osu.Framework/Configuration/BindableList.cs]: You didn't check what this does, right?

also consider using [this](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/is#a-nametype--type-pattern-a)

=>
```
if (input is IEnumerable<T> enumerable)
{
    Clear();
    underlyingList.AddRange(enumerable);
}
```

**Comment by peppy** [comment]
What's this being used for?

**Comment by miterosan** [comment]
It is being used in my private chat implementation

File: osu.Framework/Configuration/BindableList.cs [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+
+        public void Add(T item)
+        {
+            if (Disabled)
+                throw createDisabledException();
+            underlyingList.Add(item);
+            TriggerValueChange();
+        }
+
+        public void AddAll(IEnumerable<T> enumerable)
```
Comment (peppy) [inline osu.Framework/Configuration/BindableList.cs]: why not call this `AddRange` in line with .net?

File: osu.Framework/Configuration/BindableList.cs [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using osu.Framework.Lists;
+
+namespace osu.Framework.Configuration
+{
+    public class BindableList<T> : IBindable, IList<T>, IReadOnlyList<T>
+    {
+        private static InvalidOperationException createDisabledException()
```
Comment (UselessToucan) [inline osu.Framework/Configuration/BindableList.cs]: I would prefer this to be a property named `DisabledException`. I think `throw DisabledException;` looks better than `throw createDisabledException();`.

**Comment by peppy** [comment]
Are you able to push the branch this is being used in? I'm just wondering if we want to do this rather than using an existing solution like `ObservableCollection<T>`.

**Comment by miterosan** [comment]
Observablecollection does the job

**Comment by peppy** [comment]
We still may want this class though, for instance if we are saving lists to a config backing. That's why I wanted to see how it's being used first.

**Comment by miterosan** [comment]
Although I use ObservableCollection, do you still want to see the version with this class? I mean I have nothing against reimplementing the usage of that one.

**Pull request closed without merging**